### PR TITLE
*: extract a builtin function factory for date arithmetic

### PR DESCRIPTION
**Pull request by zyguan**
Hi, @hanfei1991 , is this what you mentioned in #2368? a builtin function factory for date arithmetic.

### CODE REVIEW

**Comment by hanfei1991** [comment]
@zyguan Good Job! That's just what i meaned.

File: ast/functions_test.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ func (ts *testFunctionsSuite) TestAggregateFuncExtractor(c *C) {
 
 	extractor = &AggregateFuncExtractor{}
 	expr = &FuncCallExpr{
-		FnName: model.NewCIStr("DATE_ARITH"),
+		FnName: model.NewCIStr("FAKE_FUNC"),
```
Comment (zyguan) [inline ast/functions_test.go]: By my understanding, the `expr` here is just a fake function expression for the test, no matter what name it has. Please correct me if I'm wrong :)

**Comment by shenli** [comment]
@zyguan Please resolve the conflict.

File: plan/expression_test.go [hunk 2]
```diff
@@ -366,50 +366,42 @@ func (s *testExpressionSuite) TestDateArith(c *C) {
...
-				c.Assert(nil, Equals, t.AddResult)
-			} else {
-				c.Assert(v.Kind(), Equals, types.KindMysqlTime)
-				value := v.GetMysqlTime()
-				c.Assert(value.String(), Equals, t.AddResult)
+		for _, x := range []struct {
+			fnName string
+			result interface{}
+		}{
+			{"DATE_ADD", t.AddResult},
```
Comment (hanfei1991) [inline plan/expression_test.go]: use ast.DataAdd here

Comment (zyguan) [inline plan/expression_test.go]: Ahh, I forgot I've defined these function names.

**Comment by zyguan** [comment]
Hi, @shenli , I've resolved the conflicts, PTAL.

**Comment by shenli** [comment]
LGTM
@hanfei1991 @coocood  PTAL

**Comment by hanfei1991** [comment]
LGTM

**Pull request merged**